src/main.cpp

```diff
@@ -1023,3 +1036,134 @@ static bool migrate_map_database(const GameParams &game_params, const Settings &
+	std::vector<v3s16> blocks;
+	// this will ignore the read-only db
+	db->listAllLoadableBlocks(blocks);
+	db->beginSave();
+	for (std::vector<v3s16>::const_iterator it = blocks.begin(); it != blocks.end(); ++it) {
+		if (kill) return false;
+
+		std::string data;
+		db->loadBlock(*it, &data);
+		if (data.empty()) {
+			errorstream << "Failed to load block " << PP(*it) << std::endl;
+			return false;
+		}
+
+		std::istringstream iss(data);
+		u8 ver = readU8(iss);
+		decompdata += data.size();
+
+		MapBlock mb(NULL, v3s16(0,0,0), &server);
+		mb.deSerialize(iss, ver, true);
+
+		std::ostringstream oss;
+		writeU8(oss, serialize_as_ver);
+		clock_t t = clock();
+		mb.serialize(oss, serialize_as_ver, true, -1);
+		comptime += (float)(clock() - t) / CLOCKS_PER_SEC;
+
+		db->saveBlock(*it, oss.str());
+
+		if (++count % 0xFF == 0 && time(NULL) - last_update_time >= 1) {
+			std::cerr << " Recompressed " << count << " blocks, "
+				<< (100.0 * count / blocks.size()) << "% completed.\r";
+			db->endSave();
+			db->beginSave();
+			last_update_time = time(NULL);
+		}
+	}
+	std::cerr << std::endl;
+	db->endSave();
+
+	actionstream << "############" << std::endl;
+	actionstream << "Recompressed blocks:            " << count << std::endl;
+	actionstream << "Total data (before comp.):      " << (decompdata / 1024 / 1024) << " MB" << std::endl;
+	actionstream << "Total compression CPU time:     " << comptime << "s" << std::endl;
+	actionstream << "Compression CPU time per block: " << (comptime / count * 1000 * 1000) << "us" << std::endl;
+	actionstream << "Compression speed:              " << (decompdata / comptime / 1024) << " KB/s" << std::endl;
+	actionstream << "############" << std::endl;
+	return true;
+}
+
+static bool decompresstimes_database(const GameParams &game_params, const Settings &cmd_args, const Address &addr)
+{
+	float decomptime = 0.0;
+	u32 compdata = 0;
+
+	Settings world_mt;
+	std::string world_mt_path = game_params.world_path + DIR_DELIM + "world.mt";
+	if (!world_mt.readConfigFile(world_mt_path.c_str())) {
+		errorstream << "Cannot read world.mt!" << std::endl;
+		return false;
+	}
+	std::string backend = world_mt.get("backend");
+	Server server(game_params.world_path, game_params.game_spec, false, addr, false);
+	MapDatabase *db = ServerMap::createDatabase(backend, game_params.world_path, world_mt);
+
+	u32 count = 0;
+	time_t last_update_time = 0;
+	bool &kill = *porting::signal_handler_killstatus();
+
+	// This is ok because the server doesn't actually run
+	std::vector<v3s16> blocks;
+	db->listAllLoadableBlocks(blocks);
+	for (std::vector<v3s16>::const_iterator it = blocks.begin(); it != blocks.end(); ++it) {
+		if (kill) return false;
+
+		std::string data;
+		db->loadBlock(*it, &data);
+		if (data.empty()) {
+			errorstream << "Failed to load block " << PP(*it) << std::endl;
+			return false;
+		}
+
+		std::istringstream iss(data, std::ios_base::binary);
+		compdata += data.size();
```

Comment:
...and this is flawed because not all bytes of a mapblock are compressed data

Reply:
This change now compressed the block as a whole. It should be fairly accurate.

Reply:
Sure, but that's not the case for blocks in the old format.